### DISCUSSION TITLE
关于文档中描述的竞品对比

### DISCUSSION THREAD

**Discussion by fongfai** [alibaba/formily#2432] [Q&A 2.x] [answered]
![image](https://user-images.githubusercontent.com/19684413/141688152-71e340f2-d622-40cd-aa4b-3633a0332ed3.png)

ant4.x 中也解决了值联动的精确渲染啊， 下面例子是gender联动note;  

打开独立窗口调试可看到note1这个没有被联动的字段并没有受影响
https://codesandbox.io/s/biao-dan-fang-fa-diao-yong-antd-4-17-0-alpha-9-forked-yj090?file=/index.js

**Comment by janryWang** [comment, marked as answer]
不要走值联动，值联动和值输入一个道理，你要换成显示隐藏控制，你就明白了

**Reply by fongfai** [reply]
我改成显示隐藏了 也是只渲染对应显示隐藏的字段  其他没收到影响， 不知道和你说显示隐藏的是不是同一个意思
https://codesandbox.io/s/biao-dan-fang-fa-diao-yong-antd-4-17-0-alpha-9-forked-yj090?file=/index.js

**Reply by janryWang** [reply]
没看到哪里有控制显示隐藏呀

**Reply by fongfai** [reply]
> 没看到哪里有控制显示隐藏呀

切换gender

**Reply by janryWang** [reply]
没看到效果，是不是你代码没提交

**Reply by fongfai** [reply]
> 没看到效果，是不是你代码没提交

刚不明显 我重新改了下你看下 切换gender为other时， 这个字段(customizeGender )对应控件会显示

**Reply by janryWang** [reply]
嗯。看了下，antd是后来支持了hidden，而且是4.4.0以上的版本才支持的，我在写这篇文章的时候，文档例子，还是全受控来控制显示隐藏的，只能说antd也意识到这个问题了吧。不过需要注意的是，antd是强依赖react的shouldUpdate能力做的脏检查，性能耗损是很高的，代码也很丑

不过这都还好的，还有一个问题，除了formily，其他表单方案都解决不了的问题。

就是受控更新渲染回溯问题，你可以看看designable的表单配置面板，所有输入项在输入的时候，都不会触发配置面板整树更新，但是用antd/formix/等等纯react的表单组件，基本上都是需要基于表单自身的全局value/onChange来受控更新，这个问题的本质就是，表单内部的状态，外部感知不到，只能通过最基本的value/onChange来同步状态，但是formily不一样，可以在外部维护observable value数据，下发给form，form就消费这个value就行了，内部所有精细化操作，外部都能精确感知，而不是每次都是全量刷新，所以如果antd要支持这个能力，那它必须要成为formily这种方案了

**Reply by janryWang** [reply]
还有一个点，hidden还是解决不了外部某个按钮想要主动设置某个字段的显示隐藏状态，需要全局setState才能控制，因为变化源不是来自值变化了，单纯只是一个事件，但在formily不一样，直接找到字段模型，修过它的属性即可

**Reply by fongfai** [reply]
谢谢回答的这么细致
